src/client/Microsoft.Identity.Client/Cache/CacheFallbackOperations.cs

```diff
@@ -231,105 +231,79 @@ internal static class CacheFallbackOperations
                         p.Key.ClientId.Equals(clientId, StringComparison.OrdinalIgnoreCase) &&
-                        environmentAliases.Contains(new Uri(p.Key.Authority).Host)).ToList();
+                        environmentAliases.Contains(new Uri(p.Key.Authority).Host));
+
+                bool filtered = false;
 
                 // if upn is provided then use it to filter
-                if (!string.IsNullOrEmpty(upn))
+                if (!string.IsNullOrEmpty(account?.Username))
                 {
-                    List<KeyValuePair<AdalTokenCacheKey, AdalResultWrapper>> upnEntries =
-                        listToProcess.Where(p => upn.Equals(p.Key.DisplayableId, StringComparison.OrdinalIgnoreCase)).ToList();
-
-                    listToProcess = upnEntries;
+                    listToProcess =
+                        listToProcess.Where(p => account.Username.Equals(
+                            p.Key.DisplayableId, StringComparison.OrdinalIgnoreCase));                        
+                    
+                    filtered = true;
                 }
 
                 // if uniqueId is provided then use it to filter
-                if (!string.IsNullOrEmpty(uniqueId))
+                if (!string.IsNullOrEmpty(account?.HomeAccountId?.ObjectId))
                 {
-                    List<KeyValuePair<AdalTokenCacheKey, AdalResultWrapper>> uniqueIdEntries =
-                        listToProcess.Where(p => uniqueId.Equals(p.Key.UniqueId, StringComparison.OrdinalIgnoreCase)).ToList();
+                    listToProcess =
+                        listToProcess.Where(p => account.HomeAccountId.ObjectId.Equals(
+                            p.Key.UniqueId, StringComparison.OrdinalIgnoreCase)).ToList();
 
-                    listToProcess = uniqueIdEntries;
+                    filtered = true;
                 }
 
-                List<MsalRefreshTokenCacheItem> list = new List<MsalRefreshTokenCacheItem>();
-                foreach (KeyValuePair<AdalTokenCacheKey, AdalResultWrapper> pair in listToProcess)
+                // We should filter at leasts by one criteria to ensure we adequate RT
+                if (!filtered)
                 {
-                    string homeAccountId = null;
-                    if (!string.IsNullOrEmpty(pair.Value.RawClientInfo))
-                    {
-                        homeAccountId = ClientInfo.CreateFromJson(pair.Value.RawClientInfo).ToAccountIdentifier();
-                    }
-
-                    list.Add(new MsalRefreshTokenCacheItem(
-                      new Uri(pair.Key.Authority).Host,
-                      pair.Key.ClientId,
-                      pair.Value.RefreshToken,
-                      pair.Value.RawClientInfo,
-                      familyId: null,
-                      homeAccountId: homeAccountId));
+                    logger.Warning("Could not filter ADAL entries by either UPN or unique ID, skipping.");
+                    return null;
```

Comment:
First part of fix - if no filter is presented to ADAL legacy cache (upn or object ID), do not return an RT

Reply:
now you're filtering by username and objectid right? 

---
In reply to: [503286041](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2091#discussion_r503286041) [](ancestors = 503286041)

Reply:
Yes, this is what we were doing before as well. Although it is very unlikely you can pass in an Account without username and without ID, theoretically you can do that, so I stopped this scenario. 

The actual fix is more about acknowledging the fact that ADAL saves one RT for each tenant (even though RTs are tenant-less), which was confusing. Sometime ago we started treating that as an error. It is not.